Feature/dispatch plot region aggregated

MaGering opened the pull request:
With this PR dispatch plots aggregated by regions are produced.

This PR replaces the PR in oemof-B3-methane (https://github.com/rl-institut/oemof-B3-methane/pull/28).

jnnr commented on scripts/plot_dispatch.py:
```diff
@@ -89,131 +278,42 @@ def reduce_labels(ax, simple_labels_dict):
-                ncol=4,
-                fontsize=14,
-            )
+            # Set df_stacked and df_demand_stacked to None as preparation for new cycle
+            df_stacked = None
+            df_demand_stacked = None
```
Put this in the beginning of the loop and drop the statement before the loop.

MaGering replied:
Done with commit https://github.com/rl-institut/oemof-B3/pull/242/commits/33dc7e64dd5ae60479a65c7fbe58ae89acd3b480.

jnnr commented on scripts/plot_dispatch.py:
```diff
@@ -38,6 +38,195 @@
+
+def prepare_data_for_aggregation(df_stacked, df):
+    """
+    This function concatenates stacked Dataframes
```
Actually, it concatenates a stacked DataFrame with an unstacked that will be stacked.

MaGering replied:
Done with commit https://github.com/rl-institut/oemof-B3/pull/242/commits/a5c85bcf1d6456b11c0944e9835bdbf4cfc1eec8.

jnnr reviewed (approved):
Thanks for your work! Some bits could certainly be improved
* Aggregation could happen in a function
* prepare_data_for_aggregation has some hidden logic when it receives None as first argument - not very nice,

but it works. If you find some time, you can try to improve the code. But I will not stop you from merging.

MaGering commented:
Thank you for the review @jnnr. I applied our further suggestions in commits https://github.com/rl-institut/oemof-B3/pull/242/commits/a96dd2ce8b9ac7f4d9a0a4cd709cc8c4282d297f and https://github.com/rl-institut/oemof-B3/pull/242/commits/0170d7775307fe01581ab0b09e5340f631d57daa.

Would you like to review them?

jnnr reviewed (commented):
Hi @MaGering, here are my thoughts. 
* What I meant by putting the aggregation step into a function was rather encapsulate the whole part https://github.com/rl-institut/oemof-B3/blob/a4557eb8ea42dc6cf02d69a08d8ef897468d0133/scripts/plot_dispatch.py#L283-L310 . The part that you encapsuled is not really worth a new function, imho.
* I am not super happy with https://github.com/rl-institut/oemof-B3/commit/a96dd2ce8b9ac7f4d9a0a4cd709cc8c4282d297f, either, sorry! To make it cleaner, you could collect all the dataframes in a list and concat them after finishing the loop.

MaGering commented:
Thanks for the clarification!

I reverted the unsuitable changes with commits https://github.com/rl-institut/oemof-B3/pull/242/commits/a106ffb5dd7b3dde4f6dcf251554bb2e030cfe8c and https://github.com/rl-institut/oemof-B3/pull/242/commits/7cc8ce443d6d361467d01ec645c7ac6c7cb474b6 and implemented your approaches with commits https://github.com/rl-institut/oemof-B3/pull/242/commits/45345d473dc4c6dfe8b74229ef3396793d7f528f and https://github.com/rl-institut/oemof-B3/pull/242/commits/b291aa88893478bd9a1aa34eea6c14afedcc75bc.

jnnr commented on scripts/plot_dispatch.py:
```diff
@@ -65,6 +254,50 @@ def reduce_labels(ax, simple_labels_dict):
     return handles, labels
 
 
+def aggregate_by_region(bus_files):
```
This function does more than its name suggests: It does not only aggregate, it also plots.

MaGering replied:
Here is my proposition to solve this: commit https://github.com/rl-institut/oemof-B3/pull/242/commits/b7ae6b839bc4fb62efdda99e4d10bd8c2dc164ab.

jnnr commented on scripts/plot_dispatch.py:
```diff
@@ -35,6 +35,195 @@
+        Result DataFrame with concatenated DataFrames
+
+    """
+    df_concatenated = pd.DataFrame()
+
+    for df_in_list in list_with_dfs:
+        df_concatenated = df_concatenated.append(df_in_list, ignore_index=True)
+
```
Could you do `df_concatenated = pd.concat(list_with_dfs)` instead?

MaGering replied:
Done with commit https://github.com/rl-institut/oemof-B3/pull/242/commits/6a646904352794ae67c2588b0e951a7d40caff7a.

jnnr commented on scripts/plot_dispatch.py:
```diff
@@ -35,6 +35,197 @@
+        Result DataFrame with concatenated DataFrames
+
+    """
+    df_concatenated = pd.DataFrame()
+
+    for df_in_list in list_with_dfs:
+        df_concatenated = pd.concat(
+            [df_concatenated, df_in_list], axis=0, ignore_index=True
+        )
```
How about this? Does that work as you expect it to?

MaGering replied:
Yes it does! Thank you. :)

MaGering commented:
Can I merge into dev?

jnnr commented:
> Can I merge into dev?

If it works, sure!
